proposals/GlobalUsingDirective.md

```diff
@@ -2,13 +2,194 @@
+*  **The scope of a name defined or imported by a *global_using_directive* extends over the *using_directive*s, *global_attributes* and *namespace_member_declaration*s of all the *compilation_unit*s in the program.**
+
+# Namespace and type names
+https://github.com/dotnet/csharplang/blob/master/spec/basic-concepts.md#namespace-and-type-names
+
+Changes are made to the algorithm determining the meaning of a *namespace_or_type_name* as follows.
+
+This is the relevant bullet point with proposed additions (which are **in bold**):
+*   If the *namespace_or_type_name* is of the form `I` or of the form `I<A1, ..., Ak>`:
+    * If `K` is zero and the *namespace_or_type_name* appears within a generic method declaration ([Methods](classes.md#methods)) and if that declaration includes a type parameter ([Type parameters](classes.md#type-parameters)) with name `I`, then the *namespace_or_type_name* refers to that type parameter.
+    * Otherwise, if the *namespace_or_type_name* appears within a type declaration, then for each instance type `T` ([The instance type](classes.md#the-instance-type)), starting with the instance type of that type declaration and continuing with the instance type of each enclosing class or struct declaration (if any):
+        * If `K` is zero and the declaration of `T` includes a type parameter with name `I`, then the *namespace_or_type_name* refers to that type parameter.
+        * Otherwise, if the *namespace_or_type_name* appears within the body of the type declaration, and `T` or any of its base types contain a nested accessible type having name `I` and `K` type parameters, then the *namespace_or_type_name* refers to that type constructed with the given type arguments. If there is more than one such type, the type declared within the more derived type is selected. Note that non-type members (constants, fields, methods, properties, indexers, operators, instance constructors, destructors, and static constructors) and type members with a different number of type parameters are ignored when determining the meaning of the *namespace_or_type_name*.
+    * If the previous steps were unsuccessful then, for each namespace `N`, starting with the namespace in which the *namespace_or_type_name* occurs, continuing with each enclosing namespace (if any), and ending with the global namespace, the following steps are evaluated until an entity is located:
+        * If `K` is zero and `I` is the name of a namespace in `N`, then:
+            * If the location where the *namespace_or_type_name* occurs is enclosed by a namespace declaration for `N` and the namespace declaration contains an *extern_alias_directive* or *using_alias_directive* that associates the name `I` with a namespace or type, **or any namespace declaration for `N` in the program contains a *global_using_alias_directive* that associates the name `I` with a namespace or type,** then the *namespace_or_type_name* is ambiguous and a compile-time error occurs.
```

Comment:
> Since global usings are required as the first part of a compilation unit and not inside namespaces, can this ever be true?

Since global usings are required as the first part of a compilation unit and not inside namespaces, can this ever be true?

Reply:
> Since global usings are required as the first part of a compilation unit and not inside namespaces, can this ever be true?

This is the existing way to specify the rules in a compilation unit by simply treating it as a namespace declaration. Just following the same approach instead of separating compilation units as their own thing. 

---
In reply to: [589638075](https://github.com/dotnet/csharplang/pull/4499#discussion_r589638075) [](ancestors = 589638075)

Reply:
> This is the existing way to specify the rules in a compilation unit by simply treating it as a namespace declaration.

Is that specified anywhere?

---
In reply to: [589649510](https://github.com/dotnet/csharplang/pull/4499#discussion_r589649510) [](ancestors = 589649510,589638075)

Reply:
> Is that specified anywhere?

It is specified by the fact that lookup in a compilation unit is not specified. At least I couldn't find that, if you know where it is, I'd be happy to adjust that section instead. 

---
In reply to: [589650743](https://github.com/dotnet/csharplang/pull/4499#discussion_r589650743) [](ancestors = 589650743,589649510,589638075)